core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/AttributeTypeUtils.scala

```diff
@@ -0,0 +1,297 @@
+      case _ =>
+        throw new AttributeTypeException(
+          s"not able to parse type ${fieldValue.getClass} to Integer."
+        )
+    }
+  }
+
+  @throws[AttributeTypeException]
+  def parseBoolean(fieldValue: Object): java.lang.Boolean = {
+    fieldValue match {
+      case str: String                => str.toBoolean
+      case int: Integer               => int != 0
+      case long: java.lang.Long       => long != 0
+      case double: java.lang.Double   => double != 0
+      case boolean: java.lang.Boolean => boolean
+      // Timestamp is considered to be illegal here.
+      case _ =>
+        throw new AttributeTypeException(
+          s"not able to parse type ${fieldValue.getClass} to Boolean."
+        )
+    }
+  }
+
+  @throws[AttributeTypeException]
+  def parseLong(fieldValue: Object): java.lang.Long = {
+    fieldValue match {
+      case str: String =>
+        try str.toLong
+        catch {
+          case _: NumberFormatException => parseTimestamp(fieldValue).getTime
+        }
+      case int: Integer               => int.toLong
+      case long: java.lang.Long       => long
+      case double: java.lang.Double   => double.toLong
+      case boolean: java.lang.Boolean => if (boolean) 1L else 0L
+      case timestamp: Timestamp       => timestamp.toInstant.toEpochMilli
+      case _ =>
+        throw new AttributeTypeException(s"not able to parse type ${fieldValue.getClass} to Long.")
+    }
+  }
+
+  @throws[AttributeTypeException]
+  def parseDouble(fieldValue: Object): java.lang.Double = {
+    fieldValue match {
+      case str: String                => str.toDouble
+      case int: Integer               => int.toDouble
+      case long: java.lang.Long       => long.toDouble
+      case double: java.lang.Double   => double
+      case boolean: java.lang.Boolean => if (boolean) 1 else 0
+      // Timestamp is considered to be illegal here.
+      case _ =>
+        throw new AttributeTypeException(
+          s"not able to parse type ${fieldValue.getClass} to Double."
+        )
+    }
+  }
+
+  @throws[AttributeTypeException]
+  def parseTimestamp(fieldValue: Object): Timestamp = {
+    fieldValue match {
+      case str: String =>
+        try new Timestamp(Instant.parse(str).toEpochMilli)
+        catch {
+          case _: DateTimeParseException =>
+            try Timestamp.valueOf(str)
+            catch {
+              case _: IllegalArgumentException =>
+                val utcFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+                new Timestamp(utcFormat.parse(fieldValue.toString).getTime)
+
+            }
+        }
+      case long: java.lang.Long => new Timestamp(long)
+      // Integer, Long, Double and Boolean are considered to be illegal here.
+      case _ =>
+        throw new AttributeTypeException(
+          s"not able to parse type ${fieldValue.getClass} to Timestamp."
+        )
+    }
+  }
+
+  private def tryParseInteger(fieldValue: String): AttributeType = {
```

Comment:
Since parseXXX takes java objects as input, I think tryParseXXX can also take java objects as input instead of taking strings.

Reply:
@bihaox what do you think?

Reply:
I think it is a better design. Since we previous only considered the source from the CSV file, we only used the string as input. But now we can make it more flexible

Reply:
if it does not affect the infer logic, please go ahead make the change, thanks.